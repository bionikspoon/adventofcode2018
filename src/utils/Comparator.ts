type CompareFn<T> = (l: T, r: T) => 0 | -1 | 1

export default class Comparator<T> {
  private static defaultCompareFunction<T>(l: T, r: T) {
    if (l === r) return 0

    return l < r ? -1 : 1
  }
  private compare: CompareFn<T>
  constructor(fn?: CompareFn<T>) {
    this.compare = fn || Comparator.defaultCompareFunction
  }

  public equal(l: T, r: T) {
    return this.compare(l, r) === 0
  }
  public lessThan(l: T, r: T) {
    return this.compare(l, r) < 0
  }
  public greaterThan(l: T, r: T) {
    return this.compare(l, r) > 0
  }
  public lessThanOrEqual(l: T, r: T) {
    return this.lessThan(l, r) || this.equal(l, r)
  }
  public greaterThanOrEqual(l: T, r: T) {
    return this.greaterThan(l, r) || this.equal(l, r)
  }
  public reverse() {
    const compareOriginal = this.compare
    this.compare = (l, r) => compareOriginal(r, l)
    return this
  }
}
