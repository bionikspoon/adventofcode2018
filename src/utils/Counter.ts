import R from 'ramda'

export default class Counter {
  private counts: { [key: string]: number }

  constructor() {
    this.counts = {}
  }

  public add(value: string | number) {
    const count = this.counts[value] ? this.counts[value] + 1 : 1

    this.counts = { ...this.counts, [value]: count }
    return this
  }

  public deleteValue(v: number) {
    this.counts = R.reject(R.equals(v), this.counts)

    return this
  }

  public getValue(key: string | number) {
    if (!this.counts[key]) return 0

    return this.counts[key]
  }

  public mostCommon() {
    return this.entries().sort(
      ([lKey, lCount], [rKey, rCount]) => rCount - lCount
    )
  }

  public values() {
    return Object.values(this.counts)
  }

  public entries() {
    return Object.entries(this.counts)
  }
}
