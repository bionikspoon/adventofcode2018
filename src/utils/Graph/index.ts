// Borrowed from https://github.com/trekhleb/javascript-algorithms/tree/master/src/data-structures/graph
export { default as Graph } from './Graph'
export { default as GraphEdge } from './GraphEdge'
export { default as GraphVertex } from './GraphVertex'
