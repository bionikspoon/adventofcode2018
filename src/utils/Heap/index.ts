export { default as MinHeap } from './MinHeap'
