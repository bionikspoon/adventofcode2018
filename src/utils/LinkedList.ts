import util from 'util'

export class LinkedList<T> {
  public static from<T>(xs: T[]) {
    const list = new LinkedList<T>()

    for (const x of xs) {
      list.insertEnd(new Node(x))
    }

    return list
  }

  public firstNode: Node<T> | null = null
  public lastNode: Node<T> | null = null

  public insertAfter(node: Node<T>, newNode: Node<T>) {
    newNode.prev = node
    if (node.next === null) {
      newNode.next = null
      this.lastNode = newNode
    } else {
      newNode.next = node.next
      node.next.prev = newNode
    }

    node.next = newNode
    return this
  }

  public insertBefore(node: Node<T>, newNode: Node<T>) {
    newNode.next = node
    if (node.prev === null) {
      newNode.prev = null
      this.firstNode = newNode
    } else {
      newNode.prev = node.next
      node.prev.next = newNode
    }

    node.prev = newNode

    return this
  }

  public insertBeginning(newNode: Node<T>) {
    if (this.firstNode === null) {
      this.firstNode = newNode
      this.lastNode = newNode
      newNode.prev = null
      newNode.next = null
    } else {
      this.insertBefore(this.firstNode, newNode)
    }

    return this
  }

  public insertEnd(newNode: Node<T>) {
    if (this.lastNode === null) {
      this.insertBeginning(newNode)
    } else {
      this.insertAfter(this.lastNode, newNode)
    }

    return this
  }

  public remove(node: Node<T>) {
    if (node.prev === null) {
      this.firstNode = node.next
    } else {
      node.prev.next = node.next
    }

    if (node.next === null) {
      this.lastNode = node.prev
    } else {
      node.next.prev = node.prev
    }

    return this
  }

  public toArray() {
    return Array.from(this)
  }

  public unshift(value: T) {
    this.insertBeginning(new Node(value))

    return this
  }

  public push(value: T) {
    this.insertEnd(new Node(value))

    return this
  }

  public mapNode<U>(fn: (node: Node<T>) => Node<U> | null): LinkedList<U> {
    const list = new LinkedList<U>()

    let node = this.firstNode

    while (node !== null) {
      const next = fn(node)
      if (next instanceof Node) list.insertEnd(next)
      node = node.next
    }

    return list
  }

  public mapWindow(
    fn: (window: T[], value: T) => T,
    nextAmount: number,
    prevAmount: number = 0,
    fill?: T
  ) {
    return this.mapNode(node => {
      const window = node.windowValues(nextAmount, prevAmount, fill)
      if (window.length !== nextAmount + prevAmount + 1) return null
      return new Node(fn(window, node.value))
    })
  }

  public map<U>(fn: (value: T) => U) {
    return this.mapNode((node: Node<T>): Node<U> => new Node(fn(node.value)))
  }

  public inspect(): string {
    return `LinkedList { ${this.firstNode!.join(' -> ')} }`
  }

  public [util.inspect.custom]() {
    return this.inspect()
  }

  public *[Symbol.iterator]() {
    let node = this.firstNode

    while (true) {
      if (node === null) return
      yield node.value
      node = node.next
    }
  }
}

export class Node<T> {
  public value: T

  public prev: Node<T> | null = null
  public next: Node<T> | null = null

  constructor(value: T) {
    this.value = value
  }

  public join(str: string = ' -> '): string {
    if (this.next === null) return this.inspect()

    return `${this.inspect()} ${str} ${this.next.join(str)}`
  }

  public inspect(): string {
    return `Node { ${this.value} }`
  }

  public [util.inspect.custom]() {
    return this.inspect()
  }

  public windowValues(nextAmount: number, prevAmount: number, fill?: T) {
    return [
      ...this.prevValues(prevAmount, fill),
      this.value,
      ...this.nextValues(nextAmount, fill),
    ]
  }

  public prevValues(count: number, fill?: T) {
    const results = []
    let node = this.prev

    for (let i = 0; i < count; i++) {
      if (node !== null) {
        results.unshift(node.value)
        node = node.prev
        continue
      }

      if (fill === undefined) break
      results.unshift(fill)
    }
    return results
  }

  public nextValues(count: number, fill?: T) {
    const results = []
    let node = this.next

    for (let i = 0; i < count; i++) {
      if (node !== null) {
        results.push(node.value)
        node = node.next
        continue
      }

      if (fill === undefined) break
      results.push(fill)
    }
    return results
  }
}
