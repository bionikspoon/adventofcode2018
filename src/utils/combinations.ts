export default function combinations<T>(items: T[]): Array<[T, T]> {
  return Array.from(iterateCombinations(items))
}

function* iterateCombinations<T>(items: T[]): IterableIterator<[T, T]> {
  for (let i = 0; i < items.length; i++) {
    const I = items[i]
    for (let j = i + 1; j < items.length; j++) {
      const J = items[j]
      yield [I, J]
    }
  }
}
