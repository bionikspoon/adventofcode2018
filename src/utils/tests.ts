import fs from 'fs'
import path from 'path'
import { promisify } from 'util'

const readFileAsync = promisify(fs.readFile)

export const getInput = (dirname: string, file: string) =>
  readFileAsync(path.join(dirname, 'case', file)).then(buffer =>
    buffer.toString()
  )
