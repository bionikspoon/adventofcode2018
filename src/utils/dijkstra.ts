import { Graph, GraphVertex } from './Graph'
import PriorityQueue from './PriorityQueue'

interface IDistances {
  [key: string]: number
}
export interface IDijkstraOptions<T> {
  queueCompareFn?: (
    this: PriorityQueue<T>,
    l: GraphVertex<T>,
    r: GraphVertex<T>
  ) => 0 | -1 | 1
  canTraverse?: (neighbor: GraphVertex<T>) => boolean
  shouldPoll?: (vertex: GraphVertex<T>) => boolean
}
export default function dijkstra<T>(
  graph: Graph<T>,
  startVertex: GraphVertex<T>,
  options: IDijkstraOptions<T> = {}
) {
  const visitedVertices: { [key: string]: GraphVertex<T> } = {}
  const queue = new PriorityQueue<GraphVertex<T>>(options.queueCompareFn)
  const { distances, previousVertices } = initDistances(graph)

  distances[startVertex.getKey()] = 0
  queue.add(startVertex, distances[startVertex.getKey()])

  while (!queue.isEmpty()) {
    const currentVertex = queue.poll()!

    graph.getNeighbors(currentVertex).forEach(neighbor => {
      if (visitedVertices[neighbor.getKey()]) return
      if (options.canTraverse && !options.canTraverse(neighbor)) return

      const distanceToNeighborFromCurrent =
        distances[currentVertex.getKey()] +
        graph.findEdge(currentVertex, neighbor)!.weight

      if (distanceToNeighborFromCurrent < distances[neighbor.getKey()]) {
        distances[neighbor.getKey()] = distanceToNeighborFromCurrent
        updateNeighborPriority(queue, neighbor, distances)
        previousVertices[neighbor.getKey()] = currentVertex
      }

      ensureQueueHasNeighbor(queue, neighbor, distances)
    })

    visitedVertices[currentVertex.getKey()] = currentVertex
    if (options.shouldPoll && !options.shouldPoll(currentVertex)) break
  }

  return { distances, previousVertices }
}
function initDistances<T>(graph: Graph<T>) {
  const distances: IDistances = {}
  const previousVertices: { [key: string]: GraphVertex<T> | null } = {}

  graph.getAllVertices().forEach(vertex => {
    distances[vertex.getKey()] = Infinity
    previousVertices[vertex.getKey()] = null
  })

  return { distances, previousVertices }
}

function updateNeighborPriority<T>(
  queue: PriorityQueue<GraphVertex<T>>,
  neighbor: GraphVertex<T>,
  distances: IDistances
): void {
  if (queue.hasValue(neighbor)) {
    queue.changePriority(neighbor, distances[neighbor.getKey()])
  }
}

function ensureQueueHasNeighbor<T>(
  queue: PriorityQueue<GraphVertex<T>>,
  neighbor: GraphVertex<T>,
  distances: IDistances
): void {
  if (!queue.hasValue(neighbor)) {
    queue.add(neighbor, distances[neighbor.getKey()])
  }
}
