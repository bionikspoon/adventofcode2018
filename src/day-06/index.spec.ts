import { findLargestFiniteArea, toGrid } from '.'
import { getInput } from '../utils/tests'

describe('part 1', () => {
  describe.each`
    file           | expected
    ${'input.txt'} | ${17}
  `('#findLargestFiniteArea given $file', ({ file, expected }) => {
    // ${'part-1-case-1.txt'} | ${17}
    let input: string

    beforeEach(async () => {
      input = await getInput(__dirname, file)
    })

    test('it finds the largest area that is not infinite', () => {
      expect(findLargestFiniteArea(input)).toEqual(expected)
    })

    // xtest('it matches snapshot from plotting initial points', () => {
    //   const subject = toGrid(input)

    //   expect(subject.toString()).toMatchSnapshot()
    // })

    // xtest('it matches snapshot after growing', () => {
    //   const subject = toGrid(input).growAll()

    //   expect(subject.toString()).toMatchSnapshot()
    // })
  })
})
