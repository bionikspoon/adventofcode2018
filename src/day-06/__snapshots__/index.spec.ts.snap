// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`part 1 #findLargestFiniteArea given "part-1-case-1.txt" it matches snapshot after growing 1`] = `
"00000.2222
00000.2222
0003342222
0033342222
..33344222
11.3444422
111.4444..
111.444555
111.445555
111.555555
111.555555"
`;

exports[`part 1 #findLargestFiniteArea given "part-1-case-1.txt" it matches snapshot from plotting initial points 1`] = `
"..........
.0........
..........
........2.
...3......
.....4....
.1........
..........
..........
........5.
.........."
`;
